app/controllers/submissions_controller.rb

```diff
@@ -202,6 +240,111 @@ def downloadAll
               filename: "#{@course.name}_#{@course.semester}_#{@assessment.name}_submissions.zip")
   end
 
+  action_auth_level :downloadBatch, :course_assistant
+  def downloadBatch
+    submission_ids = params[:submission_ids]
+    flash[:error] = "Cannot index submissions for nil assessment" if @assessment.nil?
+
+    unless @assessment.valid?
+      @assessment.errors.full_messages.each do |msg|
+        flash[:error] += "<br>#{msg}"
+      end
+      flash[:html_safe] = true
+    end
+
+    if @assessment.disable_handins
+      flash[:error] = "There are no submissions to download."
+      if @cud.course_assistant
+        redirect_to([@course, @assessment])
+      else
+        redirect_to([@course, @assessment, :submissions])
+      end
+      return
+    end
+
+    submissions = submission_ids.map { |sid| @assessment.submissions.find_by(id: sid) }
+    submissions = submissions.select { |s| s && @cud.can_administer?(s.course_user_datum) }
+
+    filedata = submissions.collect do |s|
+      p = s.handin_file_path
+      email = s.course_user_datum.user.email
+      [p, download_filename(p, email)] if !p.nil? && File.exist?(p) && File.readable?(p)
+    end.compact
+
+    result = Archive.create_zip filedata # result is stringIO to be sent
+    if result.nil?
+      flash[:error] = "There are no submissions to download."
+      if @cud.course_assistant
+        redirect_to([@course, @assessment])
+      else
+        redirect_to([@course, @assessment, :submissions])
+      end
+      return
+    end
+
+    send_data(result.read, # to read from stringIO object returned by create_zip
```

Comment:
I think doing this requires some javascript and css to showcase the flash since flashes dont work before send_data (https://stackoverflow.com/questions/41643886/how-to-proceed-after-send-data-from-controller), I also feel like there's the obvious sign that a file is downloaded so I'm not to sure if it makes sense

Reply:
I see, sounds good.